portal-bridge/src/types/state/execution.rs

```diff
@@ -28,72 +44,129 @@ struct GenesisConfig {
 }
 
-#[warn(dead_code)]
 pub struct State {
-    pub accounts: BTreeSet<Address>,
-    trie: Box<EthTrie<MemoryDB>>,
+    pub database: CacheDB<EmptyDB>,
     next_block_number: u64,
 }
 
 const GENESIS_STATE_FILE: &str = "resources/genesis/mainnet.json";
 
 impl State {
     pub fn new() -> anyhow::Result<Self> {
-        let mut trie = EthTrie::new(Arc::new(MemoryDB::new(false)));
+        let mut cache_db = CacheDB::new(EmptyDB::default());
         let mut accounts = BTreeSet::new();
 
         let genesis_file = File::open(GENESIS_STATE_FILE)?;
         let genesis: GenesisConfig = serde_json::from_reader(BufReader::new(genesis_file))?;
 
         for (address, alloc_balance) in genesis.alloc {
             accounts.insert(address);
-
-            let mut account = AccountState::default();
-            account.balance += alloc_balance.balance;
-            trie.insert(keccak256(address).as_ref(), &alloy_rlp::encode(account))?;
+            let mut account = DbAccount::default();
+            account.info.balance += alloc_balance.balance;
+            let account_info = account.info.clone();
+            cache_db.accounts.insert(address, account);
+            cache_db.trie.lock().insert(
+                keccak256(address).as_ref(),
+                &alloy_rlp::encode(AccountStateInfo {
+                    nonce: account_info.nonce,
+                    balance: account_info.balance,
+                    storage_root: keccak256([EMPTY_STRING_CODE]),
+                    code_hash: account_info.code_hash,
+                }),
+            )?;
         }
 
-        let root = B256::from_slice(trie.root_hash()?.as_slice());
+        let root = B256::from_slice(cache_db.trie.lock().root_hash()?.as_slice());
         ensure!(
             root == genesis.state_root,
             "Root doesn't match state root from genesis file"
         );
         Ok(State {
-            accounts,
-            trie: Box::from(trie),
             next_block_number: 1,
+            database: cache_db,
         })
     }
 
     pub fn process_block(&mut self, block_tuple: &BlockTuple) -> anyhow::Result<BTreeSet<Address>> {
+        info!(
+            "State EVM processing block {}",
+            block_tuple.header.header.number
+        );
         ensure!(
             self.next_block_number == block_tuple.header.header.number,
             "Expected block {}, received {}",
             self.next_block_number,
             block_tuple.header.header.number
         );
 
-        let mut updated_accounts = BTreeSet::new();
+        let mut updated_accounts: BTreeSet<Address> = BTreeSet::new();
+
+        // initalize evm environment
+        let mut env = Env::default();
+        env.block.number = U256::from(block_tuple.header.header.number);
+        env.block.coinbase = block_tuple.header.header.author;
+        env.block.timestamp = U256::from(block_tuple.header.header.timestamp);
+        if get_spec_id(block_tuple.header.header.number) >= SpecId::MERGE {
```

Comment:
nit: probably you want `get_spec_id(block_tuple.header.header.number).is_enabled_in(SpecId::MERGE)`